packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -230,15 +230,23 @@ class FlutterDevice {
       isWaitingForVm = true;
       vm_service.VmService service;
       if (!disableDds) {
-        await device.dds.startDartDevelopmentService(
-          observatoryUri,
-          ddsPort,
-          ipv6,
-          disableServiceAuthCodes,
-        );
+        try {
```

Comment:
Can you explain this startup process and the error cases a bit more?

Reply:
Sure. I'm happy to add this information in comments as well if you'd like.

This first case here is meant to catch errors that occur during DDS startup (e.g., failure to bind to a port, failure to connect to the VM service, attaching to a VM service with existing clients, etc).

The second try block below with the new `Future.any(...)` code handles cases where the VM service connection goes down before flutter_tools connects to DDS. The DDS `done` future completes when DDS shuts down, including after an error, so if this future completes before `connectToVmService` we know that something went wrong that caused DDS to shutdown early.

Reply:
Documented too please :)

Reply:
Done.